consensus/istanbul/backend/announce.go

```diff
@@ -269,87 +231,99 @@ func (sb *Backend) retrieveActiveAndRegisteredValidators() (map[common.Address]b
 }
 
 func (sb *Backend) handleIstAnnounce(payload []byte) error {
-	sb.logger.Trace("Handling an IstanbulAnnounce message")
-
-	msg := new(announceMessage)
+	msg := new(istanbul.Message)
 
 	// Decode message
-	err := msg.FromPayload(payload)
+	err := msg.FromPayload(payload, istanbul.GetSignatureAddress)
 	if err != nil {
 		sb.logger.Error("Error in decoding received Istanbul Announce message", "err", err, "payload", hex.EncodeToString(payload))
 		return err
 	}
 
-	// Verify message signature
-	if err := msg.VerifySig(); err != nil {
-		sb.logger.Error("Error in verifying the signature of an Istanbul Announce message", "err", err, "AnnounceMsg", msg.String())
-		return err
-	}
+	sb.logger.Trace("Handling an IstanbulAnnounce message", "from", msg.Address)
 
 	// If the message is originally from this node, then ignore it
 	if msg.Address == sb.Address() {
 		sb.logger.Trace("Received an IstanbulAnnounce message originating from this node. Ignoring it.")
 		return nil
 	}
 
+	var announceData announceData
+	err = rlp.DecodeBytes(msg.Msg, &announceData)
+	if err != nil {
+		sb.logger.Error("Error in decoding received Istanbul Announce message content", "err", err, "IstanbulMsg", msg.String())
+		return err
+	}
+
+	if view, err := sb.valEnodeTable.GetViewFromAddress(msg.Address); err == nil && announceData.View.Cmp(view) <= 0 {
+		sb.logger.Trace("Received an old announce message", "senderAddr", msg.Address, "messageView", announceData.View, "currentEntryView", view)
+		return errOldAnnounceMessage
+	}
+
 	// If the message is not within the registered validator set, then ignore it
 	regAndActiveVals, err := sb.retrieveActiveAndRegisteredValidators()
 	if err != nil {
 		return err
 	}
 
 	if !regAndActiveVals[msg.Address] {
-		sb.logger.Warn("Received an IstanbulAnnounce message from a non registered validator. Ignoring it.", "AnnounceMsg", msg.String(), "validators", regAndActiveVals, "err", err)
+		sb.logger.Warn("Received an IstanbulAnnounce message from a non registered validator. Ignoring it.", "IstanbulMsg", msg.String(), "validators", regAndActiveVals, "err", err)
 		return errUnauthorizedAnnounceMessage
 	}
 
-	// Decrypt the EnodeURL
-	nodeKey := ecies.ImportECDSA(sb.GetNodeKey())
-
-	encryptedEndpoint := []byte("")
-	destAddresses := make([]string, 0, len(msg.EncryptedEndpointData))
-	for _, entry := range msg.EncryptedEndpointData {
-		destAddresses = append(destAddresses, common.BytesToAddress(entry[0]).String())
-		if bytes.Equal(entry[0], sb.Address().Bytes()) {
-			encryptedEndpoint = entry[1]
+	var node *enode.Node
+	var destAddresses = make([]string, 0, len(announceData.AnnounceRecords))
+	for _, announceRecord := range announceData.AnnounceRecords {
```

Comment:
Geth actually already checks for a max message size in the eth message handler before it passes it to the istanbul message handler:  https://github.com/celo-org/celo-blockchain/blob/master/eth/handler.go#L367.

I'd like to say that I had the foresight of considering that issue and knew about this check, but can't admit that :slightly_smiling_face:.  Great catch!

I made changes based on your suggestion.  One thing to note is that since the full table is signed by the originating validator, the node will regossiped dup entries.  If we don't want to regossip the dups, then the originating validator will have to sign each entry individually.

Reply:
What if we just don't re-gossip a message if it has dups or irrelevant entries (or if the entry intended for us is unable to be decoded)

Reply:
Ya, that makes sense.